test/end2end_test.go

```diff
@@ -7032,6 +7032,84 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 	}()
 }
 
+// TestClientTransportRestartsAfterStreamIdExhausted tests that the client transport
+// drains and restarts when next stream ID exceeds MaxStreamID. This test also
+// verifies that subsequent RPCs use a new client transport and the old transport
+// is closed.
+func (s) TestClientTransportRestartsAfterStreamIdExhausted(t *testing.T) {
+	// overriding MaxStreamIDForTesting.
+	originalMaxStreamID := transport.MaxStreamIDForTesting
+	transport.MaxStreamIDForTesting = 5
+	defer func() {
+		transport.MaxStreamIDForTesting = originalMaxStreamID
+	}()
+
+	// setting up StubServer.
+	s := grpc.NewServer()
+	ss := &stubserver.StubServer{
+		FullDuplexCallF: func(stream testpb.TestService_FullDuplexCallServer) error {
+			for i := 0; i < 10; i++ {
+				m := &grpc.PreparedMsg{}
+				stream.SendMsg(m)
+			}
+			return nil
+		},
+	}
+	testpb.RegisterTestServiceServer(s, ss)
+
+	// setting up gRPC server with ListenerWrapper.
+	lisWrap := testutils.NewListenerWrapper(t, nil)
+	go s.Serve(lisWrap)
+	defer s.Stop()
+
+	cc, err := grpc.Dial(lisWrap.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		t.Fatalf("dial(%q): %v", lisWrap.Addr().String(), err)
+	}
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
+	defer cancel()
+
+	client := testpb.NewTestServiceClient(cc)
+
+	createStreamAndRecv := func() {
+		stream, err := client.FullDuplexCall(ctx)
+		if err != nil {
+			t.Fatalf("creating FullDuplex stream: %v", err)
+		}
+		if _, err = stream.Recv(); err != nil && err != io.EOF {
+			t.Fatalf("stream.Recv() = _, %v want: nil or EOF", err)
+		}
```

Comment:
Don't we know what we should receive?  Shouldn't it always be `nil` since the server sends 10 messages?